File: languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/Envelope/EnvelopeEncryptionBytesImplTest.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using GoDaddy.Asherah.AppEncryption.Envelope;
using Moq;
using Newtonsoft.Json.Linq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption.Envelope
{
    [Collection("Logger Fixture collection")]
    public class EnvelopeEncryptionBytesImplTest
    {
        private readonly Mock<EnvelopeEncryptionJsonImpl> envelopeEncryptionJsonImplMock;
        private readonly EnvelopeEncryptionBytesImpl envelopeEncryptionBytesImpl;

        public EnvelopeEncryptionBytesImplTest()
        {
            envelopeEncryptionJsonImplMock = new Mock<EnvelopeEncryptionJsonImpl>();
            envelopeEncryptionBytesImpl = new EnvelopeEncryptionBytesImpl(envelopeEncryptionJsonImplMock.Object);
        }

        [Fact]
        private void TestDecryptDataRowRecord()
        {
            byte[] expectedBytes = { 0, 1 };

            envelopeEncryptionJsonImplMock.Setup(x => x.DecryptDataRowRecord(It.IsAny<JObject>()))
                .Returns(expectedBytes);

            ImmutableDictionary<string, string> immutableDictionary = new Dictionary<string, string> { { "key", "value" } }.ToImmutableDictionary();
            byte[] dataRowRecordBytes = new Asherah.AppEncryption.Util.Json(JObject.FromObject(immutableDictionary)).ToUtf8();
            byte[] actualBytes = envelopeEncryptionBytesImpl.DecryptDataRowRecord(dataRowRecordBytes);
            Assert.Equal(expectedBytes, actualBytes);
        }

        [Fact]
        private void TestEncryptPayload()
        {
            ImmutableDictionary<string, string> immutableDictionary = new Dictionary<string, string> { { "key", "value" } }.ToImmutableDictionary();
            JObject dataRowRecord = JObject.FromObject(immutableDictionary);
            byte[] expectedBytes = { 123, 34, 107, 101, 121, 34, 58, 34, 118, 97, 108, 117, 101, 34, 125 };

            envelopeEncryptionJsonImplMock.Setup(x => x.EncryptPayload(It.IsAny<byte[]>())).Returns(dataRowRecord);

            byte[] actualResult = envelopeEncryptionBytesImpl.EncryptPayload(new byte[] { 0, 1 });
            Assert.Equal(expectedBytes, actualResult);
        }

        [Fact]
        private void TestDisposeSuccess()
        {
            envelopeEncryptionBytesImpl.Dispose();

            // Verify proper resources are closed
            envelopeEncryptionJsonImplMock.Verify(x => x.Dispose());
        }

        [Fact]
        private void TestDisposeWithDisposeFailShouldReturn()
        {
            envelopeEncryptionJsonImplMock.Setup(x => x.Dispose()).Throws(new SystemException());
            envelopeEncryptionBytesImpl.Dispose();

            envelopeEncryptionJsonImplMock.Verify(x => x.Dispose());
        }
    }
}

File: java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/envelope/EnvelopeEncryptionBytesImplTest.java
package com.godaddy.asherah.appencryption.envelope;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import org.json.JSONObject;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.godaddy.asherah.appencryption.utils.Json;
import com.google.common.collect.ImmutableMap;

@ExtendWith(MockitoExtension.class)
class EnvelopeEncryptionBytesImplTest {
  @Mock
  EnvelopeEncryptionJsonImpl envelopeEncryptionJsonImpl;

  @InjectMocks
  EnvelopeEncryptionBytesImpl envelopeEncryptionBytesImpl;

  @Test
  void testDecryptDataRowRecord() {
    byte[] expectedBytes = new byte[]{0, 1};
    // JSONObject doesn't play nice with Mockito, but we really just care about the base64 and utf8 operations succeeding
    when(envelopeEncryptionJsonImpl.decryptDataRowRecord(any())).thenReturn(expectedBytes);

    byte[] dataRowRecordBytes = new Json(new JSONObject(ImmutableMap.of("key", "value"))).toUtf8();
    byte[] actualBytes = envelopeEncryptionBytesImpl.decryptDataRowRecord(dataRowRecordBytes);
    assertArrayEquals(expectedBytes, actualBytes);
  }

  @Test
  void testEncryptPayload() {
    JSONObject dataRowRecord = new JSONObject(ImmutableMap.of("key", "value"));
    byte[] expectedBytes = new byte[]{123, 34, 107, 101, 121, 34, 58, 34, 118, 97, 108, 117, 101, 34, 125};
    when(envelopeEncryptionJsonImpl.encryptPayload(any())).thenReturn(dataRowRecord);

    byte[] actualResult = envelopeEncryptionBytesImpl.encryptPayload(new byte[]{0, 1});
    assertArrayEquals(expectedBytes, actualResult);
  }

  @Test
  void testCloseSuccess() {
    envelopeEncryptionBytesImpl.close();

    // Verify proper resources are closed
    verify(envelopeEncryptionJsonImpl).close();
  }

  @Test
  void testCloseWithCloseFailShouldReturn() {
    doThrow(RuntimeException.class).when(envelopeEncryptionJsonImpl).close();
    envelopeEncryptionBytesImpl.close();

    verify(envelopeEncryptionJsonImpl).close();
  }

}
